system-addon/lib/DownloadsManager.jsm

```diff
@@ -0,0 +1,183 @@
+    this.element = browser;
+    this.element._shell = this;
+  }
+
+  get download() {
+    return this._download;
+  }
+
+  get fileType() {
+    if (!this.download.target.path) {
+      return null;
+    }
+    let items = this.download.target.path.split(".");
+    return items[items.length - 1].toUpperCase();
+  }
+
+  downloadsCmd_copyLocation() {
+    let clipboard = Cc["@mozilla.org/widget/clipboardhelper;1"].getService(Ci.nsIClipboardHelper);
+    clipboard.copyString(this.download.source.url);
+  }
+
+  downloadsCmd_openReferrer() {
+    this.element.openNewTabWith(this.download.source.referrer);
+  }
+}
+
+this.DownloadsManager = class DownloadsManager {
+  constructor(store) {
+    this._downloadData = null;
+    this._window = null;
+    this._store = null;
+    this._viewableDownloadItems = new Map();
+    this._downloadTimer = null;
+  }
+
+  setTimeout(callback, delay) {
+    let timer = Cc["@mozilla.org/timer;1"].createInstance(Ci.nsITimer);
+    timer.initWithCallback(callback, delay, Ci.nsITimer.TYPE_ONE_SHOT);
+    return timer;
+  }
+
+  formatDownload(element) {
+    const downloadedItem = element.download;
+    let description;
+    if (element.fileType) {
+      // If we have a file type: '1.5 MB — PNG'
+      description = `${element.sizeStrings.stateLabel} \u2014 ${element.fileType}`;
+    } else {
+      // If we do not have a file type: '1.5 MB'
+      description = `${element.sizeStrings.stateLabel}`;
+    }
+    return {
+      hostname: new URL(downloadedItem.source.url).hostname,
+      url: downloadedItem.source.url,
+      path: downloadedItem.target.path,
+      title: element.displayName,
+      description,
+      referrer: downloadedItem.source.referrer
+    };
+  }
+
+  init(store) {
+    this._store = store;
+    this._browser = Services.appShell.hiddenDOMWindow;
+    this._downloadData = DownloadsCommon.getData(this._browser.ownerGlobal, true, false, true);
+    this._downloadData.addView(this);
+  }
+
+  onDownloadAdded(download) {
+    const elem = new DownloadElement(download, this._browser);
+    const downloadedItem = elem.download;
+    if (!this._viewableDownloadItems.has(downloadedItem.source.url)) {
+      this._viewableDownloadItems.set(downloadedItem.source.url, elem);
+
+      // Debounce incoming onDownloadAdded notifications
+      if (this._downloadTimer) {
+        this._downloadTimer.delay = DOWNLOAD_CHANGED_DELAY_TIME;
+      } else {
+        this._downloadTimer = this.setTimeout(() => {
+          this._downloadTimer = null;
+          this._store.dispatch({type: at.DOWNLOAD_CHANGED});
+        }, DOWNLOAD_CHANGED_DELAY_TIME);
+      }
+    }
+  }
+
+  onDownloadRemoved(download) {
+    if (this._viewableDownloadItems.has(download.source.url)) {
+      this._viewableDownloadItems.delete(download.source.url);
+      this._store.dispatch({type: at.DOWNLOAD_CHANGED});
+    }
+  }
+
+  async getDownloads({threshold, numItems = this._viewableDownloadItems.size, onlyExists = false}) {
```

Comment:
If `threshold` is required, probably make it as the first parameter and have the second the options object